### PR TITLE
relayburn-analyze: port replacement-savings + tool-call-patterns

### LINKED ISSUE #270: [Rust port] relayburn-analyze: replacement-savings + tool-call-patterns
**Parent:** #244
**Depends on:** #242, #266 (foundation), #268 (findings)

## Context

Two closely-coupled detectors that score "this tool call could have been replaced by something cheaper." `replacement-savings` is the primitive (token-cost estimate per call, summed per tool); `tool-call-patterns` is the higher-level finding generator that uses it to label categories like `redundant-read`, `bash-substitute-for-tool`, etc.

### `replacement_savings.rs` (port of `replacement-savings.ts`, ~154 LoC)
- Port `DEFAULT_REPLACED_TOOL_TOKEN_COST` (frozen lookup table — encode as `phf::Map` or `&'static [(...)]` literal), `DEFAULT_FALLBACK_TOKEN_COST = 800`.
- Port `ReplacementSavingsOptions`, `ToolCallSavings`, `ToolSavingsAggregate`, `ReplacementSavingsSummary` types.
- Port `estimateSavingsForToolCall(call, options)`, `summarizeReplacementSavings(calls, options)`.
- Math is straightforward token-cost lookup × count; just keep f64 ordering identical.

### `tool_call_patterns.rs` (port of `tool-call-patterns.ts`, ~479 LoC)
- Port `ToolCallPatternCategory` (string-literal union — closed enum), `ToolCallPatternFinding`, `DetectToolCallPatternsOptions`.
- Port `detectToolCallPatterns(turns, options)` and `toolCallPatternToFinding(pattern)`.
- Detector walks turn-level `ToolCall[]` and groups by category. Categories include `redundant-read`, `repeated-grep`, `bash-substitute-for-tool`, etc. — see the TS source for the full list.
- Adapter to `WasteFinding` lives in `findings.rs` (already ported in #268) — wire it through.

## Conformance gate

- `packages/analyze/src/replacement-savings.test.ts` (105 lines)
- `packages/analyze/src/tool-call-patterns.test.ts` (301 lines)

## Acceptance

- [ ] `cargo test -p relayburn-analyze` green for both modules.
- [ ] Public surface: `DEFAULT_REPLACED_TOOL_TOKEN_COST`, `DEFAULT_FALLBACK_TOKEN_COST`, `estimateSavingsForToolCall`, `summarizeReplacementSavings`, `detectToolCallPatterns`, `toolCallPatternToFinding`, plus all listed types.
- [ ] Category labels emit byte-identical strings to TS.
- [ ] Token-cost table covers the same tools as TS in the same priority order.

## Files

- `packages/analyze/src/replacement-savings.ts` + `replacement-savings.test.ts`
- `packages/analyze/src/tool-call-patterns.ts` + `tool-call-patterns.test.ts`

### PR DESCRIPTION
**Pull request by willwashburn**
Closes #270.

## Summary

Ports two coupled detectors from `@relayburn/analyze` to the Rust crate:

- **`replacement_savings.rs`** — per-call token-savings estimate + summary using the static `DEFAULT_REPLACED_TOOL_TOKEN_COST` lookup table (15 tool entries, byte-identical to TS) and `DEFAULT_FALLBACK_TOKEN_COST` (800).
- **`tool_call_patterns.rs`** — vendor-neutral pattern detector covering `search-sequence`, `edit-cluster`, `bash-git-state`, `bash-test-run`, `bash-gh-pr`. Category labels serialize to the same kebab-case strings as the TS union. Includes the `WasteFinding` adapter (`tool_call_pattern_to_finding`).

## Conformance

The full TS test suites (`replacement-savings.test.ts` + `tool-call-patterns.test.ts`) are ported verbatim as conformance gates. `cargo test -p relayburn-analyze` is green: 107 passed; 0 failed (18 new tests across the two modules).

## Notes

- `severity_from_usd` exposed via a thin `severity_from_usd_pub` wrapper in `findings.rs` so the new module shares the same severity tier table as the existing finding adapters.
- Added `phf` to `crates/relayburn-analyze/Cargo.toml` (already a workspace dep via `relayburn-reader`) for the static lookup table.
- `IndexMap` used for `ReplacementSavingsSummary.by_tool` to preserve insertion order, matching TS `Map` semantics.

## Test plan

- [x] `cargo test -p relayburn-analyze` — 107 passed, 0 failed
- [x] `cargo build --workspace` — clean
- [x] `cargo clippy -p relayburn-analyze --all-targets` — no new warnings (pre-existing reader warnings unrelated)

https://claude.ai/code/session_01JWDiud3kiyrZscm22JXuFD

---
_Generated by [Claude Code](https://claude.ai/code/session_01JWDiud3kiyrZscm22JXuFD)_

### CODE REVIEW

File: crates/relayburn-analyze/src/tool_call_patterns.rs [hunk 1]
```diff
@@ -0,0 +1,992 @@
...
+        return true;
+    }
+    if bin == "go" && parsed.subcommand.as_deref() == Some("test") {
+        return true;
+    }
+    if matches!(bin, "pnpm" | "npm" | "yarn" | "bun") {
+        if let Some(sub) = parsed.subcommand.as_deref() {
+            return sub == "test" || sub.starts_with("test:") || sub.starts_with("test ");
+        }
+    }
```
[the remark this answers is not in the document]
Comment (willwashburn) [reply crates/relayburn-analyze/src/tool_call_patterns.rs:355]: Confirmed — analysis is correct. `BashParse.subcommand` for package runners is always a single token (the `run X` case in `classifier.rs:419-428` replaces, doesn't concatenate, and `TWO_PART_SUBCOMMANDS` only covers `docker`/`gh`/`go`/`uv`), so `sub.starts_with("test ")` is unreachable.

Not a port-translation slip though — the same dead branch exists verbatim in the TS source at `packages/analyze/src/tool-call-patterns.ts:348`. Since this PR's contract is byte-identical fidelity to TS (the test suites are ported verbatim as conformance gates), I'm leaving the line alone here. Filing a follow-up to drop it from both `tool_call_patterns.rs` and `tool-call-patterns.ts` in lockstep so the Rust ↔ TS conformance corpus stays aligned.

[the remark this answers is not in the document]
Comment (willwashburn) [reply crates/relayburn-analyze/src/tool_call_patterns.rs:355]: Fixed in 4844363 — dropped the dead branch from both `tool_call_patterns.rs` and `tool-call-patterns.ts` in this PR rather than spinning out a follow-up. Tests still green on both sides.

File: crates/relayburn-analyze/src/replacement_savings.rs [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+    let fallback = resolve_fallback(options);
+    let overrides = options.and_then(|o| o.cost_per_call.as_ref());
+    let avg = average_replaced_cost(replaced, overrides, fallback);
+    // When `collapsedCalls` is missing but `replaces` is present, treat the
+    // call as having replaced one of each named tool. Conservative floor.
+    let calls = if collapsed > 0 { collapsed } else { replaced.len() as u64 };
+    Some(ToolCallSavings {
+        collapsed_calls: calls,
+        replaced_tools: replaced.to_vec(),
+        estimated_tokens_saved: round_to_u64(calls as f64 * avg),
```
[the remark this answers is not in the document]
Comment (willwashburn) [reply crates/relayburn-analyze/src/replacement_savings.rs:141]: Verified — declining this change. The Rust port mirrors the TS source byte-for-byte:

```ts
// packages/analyze/src/replacement-savings.ts:112-120
const collapsedCalls = call.collapsedCalls ?? 0;
const replaced = call.replacedTools ?? [];
if (collapsedCalls <= 0 && replaced.length === 0) return undefined;
…
const calls = collapsedCalls > 0 ? collapsedCalls : replaced.length;
```

TS uses `??` to fold `undefined` and `0` together, then `<= 0` and `> 0` to fall through to `replaced.length`. The Rust unwraps `None → 0` and uses the same `== 0` / `> 0` shape. Adopting the suggestion would diverge the Rust port from the TS source for an undocumented edge case (`_meta.collapsedCalls: 0` emitted alongside `_meta.replaces: [...]` — a contradictory annotation no producer should write).

The PR's contract is byte-identical fidelity to TS; the test suites are ported verbatim as conformance gates. The documented intent (TS line 117-119: "conservative floor: at least one vanilla call per listed name") is what the current code embodies. No follow-up filed — unlike the dead `test ` branch, this isn't a slip in the original; it's the intended behavior.
